teamcity: failed test: TestReplicaReproposalWithNewLeaseIndexError

The following tests appear to have failed on master (testrace): TestReplicaReproposalWithNewLeaseIndexError

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestReplicaReproposalWithNewLeaseIndexError).

[#1444709](https://teamcity.cockroachdb.com/viewLog.html?buildId=1444709):


```
TestReplicaReproposalWithNewLeaseIndexError
...o:307
sync.runtime_notifyListWait(0xc0009ea910, 0x1c)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc0009ea900)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc0010cc120, 0x5b77120, 0xc001733e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5b77120, 0xc001733e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc001dec1b0, 0xc00103c960, 0xc001dec190)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 27504 [sync.Cond.Wait]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:307
sync.runtime_notifyListWait(0xc0009ea910, 0x39)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc0009ea900)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc0010cc120, 0x5b77120, 0xc001dd55f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5b77120, 0xc001dd55f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc001dec8a0, 0xc00103c960, 0xc001dec890)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 27460 [sync.Cond.Wait]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:307
sync.runtime_notifyListWait(0xc0009ea910, 0xd)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc0009ea900)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc0010cc120, 0x5b77120, 0xc001733890)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5b77120, 0xc001733890)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc000339f60, 0xc00103c960, 0xc000339f50)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 27527 [select]:
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processLoop.func1(0x5b77120, 0xc001e205d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:748 +0x1ef
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc001e93aa0, 0xc00103c960, 0xc00033ff40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 27523 [select]:
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processLoop.func1(0x5b77120, 0xc001e20570)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:748 +0x1ef
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc001e93a50, 0xc00103c960, 0xc00033fec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4




```

Please assign, take a look and update the issue accordingly.

```
Test ended in panic.

------- Stdout: -------
W190819 14:48:15.531666 27451 storage/replica_application_result.go:219  [s1,r1/1:/M{in-ax}] failed to repropose with new lease index: boom
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1f45dcb]

goroutine 27451 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00103c960, 0x5b77120, 0xc001dd4c30)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x155
panic(0x482bd80, 0x808f640)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage.(*replicaProposer).registerProposalLocked(0xc001246d00, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal_buf.go:649 +0x4b
github.com/cockroachdb/cockroach/pkg/storage.(*propBuf).FlushLockedWithRaftGroup(0xc0012472a0, 0xc001ddb4c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal_buf.go:406 +0x35b
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked.func1(0xc001ddb4c0, 0x18, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:456 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked.func1(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:1337
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc001246d00, 0x5b77101, 0xc0011d74d8, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:1338 +0x121
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc001246d00, 0x5b77120, 0xc001e21560, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:455 +0x1e2
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady(0xc001246d00, 0x5b77120, 0xc001e21560, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:432 +0x1a0
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady(0xc000647500, 0x5b77120, 0xc001dd4c30, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3675 +0x1a0
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc0010cc120, 0x5b77120, 0xc001dd4c30)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:227 +0x33f
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5b77120, 0xc001dd4c30)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc000339d10, 0xc00103c960, 0xc000339d00)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4
```

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -76,7 +76,7 @@ type ProposalData struct {
 	// containing the command ID.
 	encodedCommand []byte
 
-	// quotaAlloc is the allocation retreived from the proposalQuota.
+	// quotaAlloc is the allocation retrieved from the proposalQuota.
 	// Once a proposal has been passed to raft modifying this field requires
 	// holding the raftMu.
 	quotaAlloc *quotapool.IntAlloc

--- a/pkg/storage/replica_proposal_buf.go
+++ b/pkg/storage/replica_proposal_buf.go
@@ -400,6 +400,11 @@ func (b *propBuf) FlushLockedWithRaftGroup(raftGroup *raft.RawNode) error {
 	// stop trying to propose commands to raftGroup.
 	var firstErr error
 	for i, p := range buf {
+		if p == nil {
+			// If we run into an error during proposal insertion, we may have reserved
+			// an array index without actually inserting a proposal.
+			continue
+		}
 		buf[i] = nil // clear buffer
 
 		// Raft processing bookkeeping.

--- a/pkg/storage/replica_proposal_buf_test.go
+++ b/pkg/storage/replica_proposal_buf_test.go
@@ -21,6 +21,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.etcd.io/etcd/raft"
 	"golang.org/x/sync/errgroup"
@@ -257,4 +258,75 @@ func TestProposalBufferRegistersAllOnProposalError(t *testing.T) {
 	err := b.flushLocked()
 	require.Equal(t, propErr, err)
 	require.Equal(t, num, p.registered)
+}
+
+// TestProposalBufferRegistrationWithInsertionErrors tests that if during
+// proposal insertion we reserve array indexes but are unable to actually insert
+// them due to errors, we simply ignore said indexes when flushing proposals.
+func TestProposalBufferRegistrationWithInsertionErrors(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	var p testProposer
+	var b propBuf
+	b.Init(&p)
+
+	num := propBufArrayMinSize / 2
+	for i := 0; i < num; i++ {
+		pd, data := newPropData(i%2 == 0)
+		_, err := b.Insert(pd, data)
+		require.Nil(t, err)
+	}
+
+	var insertErr = errors.New("failed insertion")
+	b.testing.leaseIndexFilter = func(*ProposalData) (indexOverride uint64, err error) {
+		return 0, insertErr
+	}
+
+	for i := 0; i < num; i++ {
+		pd, data := newPropData(i%2 == 0)
+		_, err := b.Insert(pd, data)
+		require.Equal(t, insertErr, err)
+	}
+	require.Equal(t, 2*num, b.Len())
+
+	require.Nil(t, b.flushLocked())
+
+	require.Equal(t, 0, b.Len())
+	require.Equal(t, num, p.registered)
+}
+
+// TestPropBufCnt tests the basic behavior of the counter maintained by the
+// proposal buffer.
+func TestPropBufCnt(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	var count propBufCnt
+	const numReqs = 10
+
+	reqLeaseInc := makePropBufCntReq(true)
+	reqLeaseNoInc := makePropBufCntReq(false)
+
+	for i := 0; i < numReqs; i++ {
+		count.update(reqLeaseInc)
+	}
+
+	res := count.read()
+	assert.Equal(t, numReqs, res.arrayLen())
+	assert.Equal(t, numReqs-1, res.arrayIndex())
+	assert.Equal(t, uint64(numReqs), res.leaseIndexOffset())
+
+	for i := 0; i < numReqs; i++ {
+		count.update(reqLeaseNoInc)
+	}
+
+	res = count.read()
+	assert.Equal(t, 2*numReqs, res.arrayLen())
+	assert.Equal(t, (2*numReqs)-1, res.arrayIndex())
+	assert.Equal(t, uint64(numReqs), res.leaseIndexOffset())
+
+	count.clear()
+	res = count.read()
+	assert.Equal(t, 0, res.arrayLen())
+	assert.Equal(t, -1, res.arrayIndex())
+	assert.Equal(t, uint64(0), res.leaseIndexOffset())
 }

